src/webview/html/messageAsHtml.js

```diff
@@ -72,14 +79,98 @@ $!${messageReactionListAsHtml(reactions, ownUser.user_id, allImageEmojiById)}
 `;
 };
 
-const widgetBody = (message: Message | Outbox) => template`
+/**
+ * Render the body of a message that has submessages.
+ *
+ * Must not be called on a message without any submessages.
+ */
+const widgetBody = (message: Message, ownUserId: UserId) => {
+  invariant(
+    message.submessages !== undefined && message.submessages.length > 0,
+    'should have submessages',
+  );
+
+  const widgetSubmessages: Array<{
+    sender_id: number,
+    content: SubmessageData,
+    ...
+  }> = message.submessages
+    .concat()
+    .sort((m1, m2) => m1.id - m2.id)
+    .filter(submessage => submessage.msg_type === 'widget')
+    .map(submessage => ({
+      sender_id: submessage.sender_id,
+      content: JSON.parse(submessage.content),
+    }));
+
+  const errorMessage = template`
 $!${message.content}
 <div class="special-message"
  ><p>Interactive message</p
  ><p>To use, open on web or desktop</p
 ></div>
 `;
 
+  const pollWidget = widgetSubmessages.shift();
+  if (!pollWidget || !pollWidget.content || pollWidget.content.widget_type !== 'poll') {
+    return errorMessage;
+  }
+
+  // $FlowFixMe[incompatible-type]
+  const pollWidgetContent: WidgetData = pollWidget.content;
+
+  if (pollWidgetContent.widget_type !== 'poll') {
+    return errorMessage;
+  }
+
+  if (pollWidgetContent.extra_data == null) {
+    // We don't expect this to happen in general, but there are some malformed
+    // messages lying around that will trigger this.
+    return template`$!${message.content}`;
+  }
```

Comment:
So from the webapp logic I quoted at https://github.com/zulip/zulip-mobile/pull/4704#discussion_r648546322 , it looks like the webapp's behavior in this case should be more like:
```js
  const { extra_data = {} } = pollWidgetContent;
```

We can vary from that, but I'd like to be explicit about doing so, because that makes it possible to compare the two implementations and spot discrepancies.

I think you showed me on your screen the other day the old message you had in mind, in the webapp, and it showed up with a behavior more like this, with just some text like "/poll". (There was an error in the JS console, suggesting the code may have errored out before it reached the code I quoted.) I think you may have suggested some reasoning for why showing the raw content like this is preferable to giving `extra_data` a default, like: at least then it'll say "/poll" instead of just looking empty.

Would you put that reasoning, and a link to that message, into the commit? And in the code, a comment saying this is different from web's behavior in this error case but we think this is better.

Reply:
I commented this case, filed #19145, and added a link to the message. I didn't modify the commit message, since it seems redundant to do that given the comment.

I tried to word it carefully, since the behaviour of this code is identical to the behaviour of the webapp as a whole, except for the error message printed to the console. The *local* behaviour is different, so it seems worth commenting still.